ghost/recommendations/src/RecommendationController.ts

```diff
@@ -0,0 +1,149 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import {Recommendation} from "./Recommendation";
+import {RecommendationService} from "./RecommendationService";
+import errors from '@tryghost/errors';
+
+type Frame = {
+    data: any,
+    options: any,
+    user: any
+};
+
+function validateString(object: any, key: string, {required = true} = {}): string|undefined {
+    if (typeof object !== 'object' || object === null) {
+        throw new errors.BadRequestError({message: `${key} must be an object`});
+    }
+
+    if (object[key] !== undefined) {
+        if (typeof object[key] !== "string") {
+            throw new errors.BadRequestError({message: `${key} must be a string`});
+        }
+        return object[key];
+    } else if (required) {
+        throw new errors.BadRequestError({message: `${key} is required`});
+    }
+}
+
+function validateBoolean(object: any, key: string, {required = true} = {}): boolean|undefined {
+    if (typeof object !== 'object' || object === null) {
+        throw new errors.BadRequestError({message: `${key} must be an object`});
+    }
+    if (object[key] !== undefined) {
+        if (typeof object[key] !== "boolean") {
+            throw new errors.BadRequestError({message: `${key} must be a boolean`});
+        }
+        return object[key];
+    } else if (required) {
+        throw new errors.BadRequestError({message: `${key} is required`});
+    }
+}
+
+export class RecommendationController {
+    service: RecommendationService;
+
+    constructor(deps: {service: RecommendationService}) {
+        this.service = deps.service;
+    }
+
+    #getFrameId(frame: Frame): string {
+        if (!frame.options) {
+            throw new errors.BadRequestError();
+        }
+
+        const id = frame.options.id;
+        if (!id) {
+            throw new errors.BadRequestError();
+        }
+
+        return id;
+    }
+
+    #getFrameRecommendation(frame: Frame): Recommendation {
+        if (!frame.data || !frame.data.recommendations || !frame.data.recommendations[0]) {
+            throw new errors.BadRequestError();
+        }
+
+        const recommendation = frame.data.recommendations[0];
+
+        const cleanedRecommendation: Omit<Recommendation, 'id'|'createdAt'> = {
+            title: validateString(recommendation, "title") ?? '',
+            url: validateString(recommendation, "url") ?? '',
```

Comment:
Required is set to `true` by default, so it will throw an error in that case. It is just that the return type of `validateString` in TypeScript returns an optional string. We could fix it by making the type of validateString a bit complex by some how checking the options that are passed to it, but I think that is a bit too much here. Technically we'll never execute the `?? ''` here.

Reply:
We might want to validate the URL a bit better in the future, but I think that is out of scope for this PR and should be added in a separate PR if needed in the future